include/flux/op/set_algorithms.hpp

```diff
@@ -0,0 +1,151 @@
+
+// Copyright (c) 2023 Jiri Nytra (jiri.nytra at gmail dot com)
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef FLUX_OP_SET_ALGORITHMS_HPP_INCLUDED
+#define FLUX_OP_SET_ALGORITHMS_HPP_INCLUDED
+
+#include <flux/core.hpp>
+
+#include <utility>
+
+namespace flux::detail {
+
+template <sequence Base1, sequence Base2, typename Cmp>
+struct set_union_adaptor
+    : flux::inline_sequence_base<set_union_adaptor<Base1, Base2, Cmp>>
+{
+private:
+    Base1 base1_;
+    Base2 base2_;
+    Cmp cmp_;
+
+public:
+    constexpr set_union_adaptor(decays_to<Base1> auto&& base1, decays_to<Base2> auto&& base2, Cmp cmp)
+        : base1_(std::move(base1)),
+          base2_(std::move(base2)),
+          cmp_(cmp)
+    {}
+
+    struct flux_sequence_traits {
+    private:
+
+        struct cursor_type {
+            cursor_t<Base1> base1_cursor;
+            cursor_t<Base2> base2_cursor;
+            enum {FIRST, SECOND} active_ = FIRST;
+        };
+
+    public:
+        using value_type = value_t<Base1>;
+
+        inline static constexpr bool is_infinite = flux::infinite_sequence<Base1> or flux::infinite_sequence<Base2>;
+
+        template <typename Self>
+        static constexpr void update(Self& self, cursor_type& cur) {
+            if (flux::is_last(self.base1_, cur.base1_cursor)) {
+                cur.active_ = cursor_type::SECOND;
+                return;
+            }
+
+            if (flux::is_last(self.base2_, cur.base2_cursor)) {
+                cur.active_ = cursor_type::FIRST;
+                return;
+            }
+
+            if (std::invoke(self.cmp_, flux::read_at(self.base2_, cur.base2_cursor), 
+                                       flux::read_at(self.base1_, cur.base1_cursor))) {
+                cur.active_ = cursor_type::SECOND;
+                return;
+            }
+
+            if (not std::invoke(self.cmp_, flux::read_at(self.base1_, cur.base1_cursor), 
+                                           flux::read_at(self.base2_, cur.base2_cursor))) {
+                flux::inc(self.base2_, cur.base2_cursor);
+            }
+
+            cur.active_ = cursor_type::FIRST;
+        }
+
+        template <typename Self>
+        static constexpr auto first(Self& self) -> cursor_type
```

Comment:
One thing I didn't mention in the tutorial (because it's quite confusing!) is how to handle const iteration for adaptors.

Not all sequences are const-iterable (single-pass sequences never are, for example), but we'd like our adaptor to support const iteration when both of its base sequences do -- and you've done that! 

But we also need to make sure we handle things the other way around as well: if either of `Base1` or `Base2` is *not* const-iterable, then the adaptor shouldn't be either. To achieve this, we need to constrain the implementation functions (first/is_last/etc) somehow when called on a `const set_union_adaptor`. There are a few different ways to do this, but in this case it's a bit more tricky because we have two underlying sequences we need to worry about, whereas most adaptors only have one.

What we actually want is a constraint that says

> If `Self` is a `const set_union_adaptor`, then require `sequence<const Base1>` and `sequence<const Base2>` to both be true

The easiest way to say this I think would be to add the following in the `private` section of the `flux_sequence_traits` struct:

```cpp
template <typename Self>
static inline constexpr bool can_const_iterate =
    std::is_const_v<Self> ?
        (flux::sequence<Base1 const> && flux::sequence<Base2 const>) : true;
```

and then constrain the functions like this:

```suggestion
        template <typename Self>
            requires can_const_iterate<Self>
        static constexpr auto first(Self& self) -> cursor_type
```

(and similarly for `is_last`, `inc` and `read_at`)

Phew! I realise that's all a bit confusing and probably not very well explained so if you have any questions please ask me :)

Reply:
Actually, now I think about it, `can_const_iterate` isn't a great name for this.... perhaps `maybe_const_iterable` might be better?

Reply:
Thanks for comprehensive explanation. I think, I understand. I changed it, including unit tests, and it works. According to the concept name, I still try to find better name but for now I used `maybe_const_iterable`.